### ISSUE TITLE
detect counts in empty lists

### ISSUE THREAD

**Issue by ghost** [makumba/makumba#21]
**Reported by @cristianbogdan on 20 Oct 2002 16:35 UTC**
<mak:list from... where ... countvar="c" />

should detect the fact that a count is made, and transform to the equivalent
<% int c; %>
<mak:list .....>
   <mak:value expr="count(x)" var="c1" />
   <%c=((Integer)c1).intValue();%>
</mak:list>

which is much faster, as counting is done in sql not in java

Migrated-From: http://trac.makumba.org/ticket/25

**Comment by ghost** [comment]
**Comment by @frederikhabils on 29 Jun 2003 22:37 UTC**
ok, I guess when implementation time comes, it would have been noticed that the
correct thing to do is closer to the following:

<% Integer c = null; %>
<mak:list .....>
   <mak:value expr="count(*)" var="c1" />
   <% c = (Integer)c1; %>
</mak:list>
<% if (c==null) c = new Integer(0); %>

(that is: Integer type + cover the case of Zero)


I think it would be nice to fix this in the near future. It's a pain in the *ss
having to type this many lines of code if you want to do performant counting. 

And it's hard to expect beginning mak-users to do this. 

Cfr e.g. http://www.best.eu.org/index.jspx#336

...which is downloading all 9000 best.johnny.Application for year 2002 on every
homepage access. :-( 

Actually, that page is dynamically counting 5 numbers ("in 2002...") that will
never change, go figure?

**Comment by ghost** [comment]
**Comment by @cristianbogdan on 30 Jun 2003 11:29 UTC**
is it true that in 99% of the cases there is a "count display" just after the
counting mak:list?

then, should we have
<mak:count from=... etc/> 
which would be implemented by a subclass of mak:list that would simply add "nil"
as a projection and display the count... that can accept a var=... afterwards

should <mak:list countVar= /> become illegal/deprecated afterwards?

**Comment by ghost** [comment]
**Comment by @frederikhabils on 30 Jun 2003 11:58 UTC**
> is it true that in 99% of the cases there is a "count display" just after the
> counting mak:list?

if your question is: "is 99% of the counting done only to display that result"
then I think the answer is no. 

I believe (a rough estimate) at least half the cases, that number is relevant
for some other presentation logic. Decide on layout (color), decide on (not)
showing certain part of text, etc.

For me, I can live with an additional mak:count. It may be more natural to the
mak:user. 

More as a personal reflex, I prefer there is only one way of doing things. But
with <mak:list maxCountVar /> not allowing to display the number without
scriptlet, I can accept the new mak:count.

I don't see any other use for the single Tag <mak:list /> other than counting,
so yeah, we may deprecate that.

**Comment by ghost** [comment]
**Comment by @cristianbogdan on 4 Jan 2005 03:06 UTC**
statistics start to become important and we need some design decissions. please
feedback!

----------
this has been discussed in various forums for a while now. i think there is a
rather important number of counts in karamba performed via closed <mak:list/>
followed by maxCountVar or mak:lastCount()

(note: there is a very-very-old idea to have closed <mak:list/> do the
correspondent of an autoform (show all fields in mak:value, see bug 781). i
guess that's sort of dead now)

so the refactoring at comment 0 still looks like a good idea. i call it
refactoring because it has to make sure that all existing <mak:list /> will work
_as_before_, but will be much more efficient.

_however_ just computing count(label) internally instead of looping like crazy
is _not_ enough. in the case where the <mak:list/> is inside another <mak:list>,
then there has to be a hidden groupBy a label from the FROM of the outer list! i
think any label will do, but i'm not sure...

-------------

so i come and ask: is this really a good idea? it may lead to hidden groupBys,
and we know that groupBy is something that people should learn, not have it done
by makumba behind the scenes.

on the other hand i agree that a shorter form of the counting should exist. and
since we do it, we can design also for other statistix (avg, sum)

could simply be

<mak:value expr="count(label)" from="T label" where="..."..../>

that is, we add from, where and groupBy to mak:value and gain statistics for it,
but keep other attributes (var, printVar, etc)

only problem i see is that if you want to do two statistics at once (e.g.
counting and average), you might be tempted to do two mak:values instead of the
twice as fast mak:list with 2 internal mak:values (it's twice as fast because
the joins are done only once in the SQL engine). but the copying of the from=,
where=, etc will probably 'shout' to the programmer that he/she is not doing the
best code.

**Comment by ghost** [comment]
**Comment by Lode.spruyt@BEST.eu.org on 4 Jan 2005 11:22 UTC**
Agreed on the fact stats are becoming important, at least in minerva we're
facing new challenges.

-------------

> rather important number of counts in karamba performed via closed <mak:list/>

absolutely.

-------------
I think the hidden groupBy thing is getting it too complicated. For an average
mak:user(as myself), it might be not that easy to read/understand code upon
reading other peoples pages.And as you say, we/they should learn it anyway.

-------------
It would be nice indeed to have a shorter form doing statistics.

>  <mak:value expr="count(label)" from="T label" where="..."..../>

Seems like a nice solution to me, compact and understandable. It leaves the
possibility to store the value in a variable using var=..., right ?

From my point of view, it would save me a _lot_ of code in minerva pages.

**Comment by ghost** [comment]
**Comment by @cristianbogdan on 4 Jan 2005 20:46 UTC**
comment 4:
> _however_ just computing count(label) internally instead of looping like crazy
> is _not_ enough. in the case where the <mak:list/> is inside another <mak:list>,
> then there has to be a hidden groupBy a label from the FROM of the outer list! i
> think any label will do, but i'm not sure...

from a discussion we're having about a certain groupBy
http://parade.best.eu.org/gwen-k/general/survey/user/viewStatistics.jspx#69

it seems that the hypothetical hidden groupBy would have to contain _all_ labels
from the enclosing mak:lists. so the groupBy= of the hypothetically transformed
mak:list would be quite easy to determine. 

so automatic transformation of
<mak:list .... maxCountVar="x"/>
into
<c:set var="x" value="0" />
<mak:list groupBy="all labels of enclosing mak:lists">
   <mak:value expr="count(*)" var="x"/>
</mak:list>  <%-- also setting mak:lastCount() to x --%>

is really easily doable, independently of the new design
<mak:value expr="whatever statistic()" from="..." where="..." var="..."
printVar="..." />

btw, i finally realized why Fred uses count(*) in comment 1. count(*) is  faster
(and really equivalent with <mak:list maxCountVar=.../>) because count(label)
returns the count of all records that have label <> null, which is maximum
count(*) but possibly smaller.

http://dev.mysql.com/doc/mysql/en/GROUP-BY-Functions.html

**Comment by ghost** [comment]
**Comment by @rudolfmayer on 4 Jan 2005 21:09 UTC**
> this has been discussed in various forums for a while now. i think there is a
> rather important number of counts in karamba performed via closed <mak:list/>
> followed by maxCountVar or mak:lastCount()


it is quite a huge number of these mak:lists. i do not exactly like it very
much, cause at least for myself when i see a mak:list, i always expect a tag body.
so, it would be nice to replace this somehow, for sure also for the performance
issues mentioned in comment 0.
 
> (note: there is a very-very-old idea to have closed <mak:list/> do the
> correspondent of an autoform (show all fields in mak:value, see bug 781). i
> guess that's sort of dead now)


i have to say, i _do_ hope that this idea is dead, cause using a mak:list for an
automatically generated form is for sure a confusing thing..
 
> so the refactoring at comment 0 still looks like a good idea. i call it
> refactoring because it has to make sure that all existing <mak:list /> will work
> _as_before_, but will be much more efficient.
> 
> _however_ just computing count(label) internally instead of looping like crazy
> is _not_ enough. in the case where the <mak:list/> is inside another <mak:list>,
> then there has to be a hidden groupBy a label from the FROM of the outer list! i
> think any label will do, but i'm not sure...


hm.. what about the case that you have two nested mak:list, but the outer
mak:list does not only contain the inner mak:list, but also does some things
(displaying results, printing some count, whatever), will a hidden group-by
added to that outer mak:list affect it's behaviour?

> so i come and ask: is this really a good idea? it may lead to hidden groupBys,
> and we know that groupBy is something that people should learn, not have it done
> by makumba behind the scenes.


agreed, groupy-by is a quite vital thing to understand if you want to understand
sql. also i am in favour of not having too many things going on behind the scenes.

then, i think that it should not be a primary target to optimise existing
counting-only mak:lists, it is easy enough to detect lists without a body, and
then they should be replaced, and maybe this can be deprecated and not supported
anymore in the future.
so, i'm in favour of a new thing, which does not care about existing code, as
long as the existing code still works.

> on the other hand i agree that a shorter form of the counting should exist. and
> since we do it, we can design also for other statistix (avg, sum)
> 
> could simply be
> 
> <mak:value expr="count(label)" from="T label" where="..."..../>
> 
> that is, we add from, where and groupBy to mak:value and gain statistics for it,
> but keep other attributes (var, printVar, etc)


this would look quite nice and handy, BUT, a big BUT, i do not like to see
mak:value used to do two rather different things.
if we are going to add the from&where to mak:value just to do statistics, i
would favour to have a mak:stat tag that is especially dedicated to that.
would be then <mak:stat expr="" from... where... var... /> [seems not
needed as you do not have pointers, and Integer/Double are seemlesly convertable
to string](printvar)
 
> only problem i see is that if you want to do two statistics at once (e.g.
> counting and average), you might be tempted to do two mak:values instead of the
> twice as fast mak:list with 2 internal mak:values (it's twice as fast because
> the joins are done only once in the SQL engine). but the copying of the from=,
> where=, etc will probably 'shout' to the programmer that he/she is not doing the
> best code.


i'd not count on the programmers thinking like that :-/ having seen the amount
of copy-paste code i did, i think copy-paste is exactly what is going to happen.

two things to avoid that situation: 
i guess there could be a way to detect this at page analysis (two mak:values
having the same from&where within the same "level" of nesting, and if this case
is detected, we could print a warning to the programmer.

on the other hand, if you'd implement a mak:stat tag, you could maybe provide
sum, avg and count to be accessible by default.
so, maybe something like this:
<mak:stat field=".." from="....." where... /> // would do the query, and store
sum(field), avg(field), count(field)
then, getting count with mak:count() [lastCount(), careful to not have a
collision with the thing for mak:lists](or)
getting avg with mak:average()
getting sum with mak:sum()

but maybe that idea is also total bollocks :-)

**Comment by ghost** [comment]
**Comment by @cristianbogdan on 4 Jan 2005 22:03 UTC**
maybe i should mention that while the <mak:list/> -> count(*) transformation
doesn't save much time for an individual page, i think it's important to
consider how much it gives to overall server time and workload.

another issue here is that for easy (automatic or manual) code conversion
mak:lastCount() could be extended to mean "count of last <mak:list > or last
count(*)"

> hm.. what about the case that you have two nested mak:list, but the outer
> mak:list does not only contain the inner mak:list, but also does some things
> (displaying results, printing some count, whatever), will a hidden group-by
> added to that outer mak:list affect it's behaviour?

no, the groupBy is added to the _inner_ mak:list, not to the outer. it doesn't
affect the outer. the groupBy just uses labels from the outer.

> then, i think that it should not be a primary target to optimise existing
> counting-only mak:lists, it is easy enough to detect lists without a body, and
> then they should be replaced, and maybe this can be deprecated and not supported
> anymore in the future.

you can imagine that it's easier for me to let others do a replacement in JSP
source instead of implementing new features of mak:list. however, without page
analysis, that replacement cannot be automatic, due to the groupBy thing.

(here are a bunch of other replacements that need be done in karamba source to
take advantage of new features, i made a list of these from 0.5.5 to 0.5.11,
sent to mak-user@  July 2004, maybe we should update it)

one other solution is to simply say: 

"<mak:list /> is illegal. if the JSP file is more than 10 days old, honour it
(maybe with the count() optimization). If the JSP file is recently developed,
show an error and indicate usage of <mak:list> <mak:value
expr="count(*)"/><mak:list>". 

only problem i see with this is that if you change a <br> to a <p> in the page
you're stuck with a count(*) and groupBy :))). another solution is to say

"<mak:list /> is illegal. if the JSP file is more than 10 days old, honour it
(maybe with the count() optimization). If the JSP file is recently developed,
transform automatically <mak:list> <mak:value expr="count(*)"/><mak:list> in the
JSP file."

about mak:stat: i like the idea, it is indeed more clear. i suggested mak:value
because it returns one single value (one row, one column).

> i guess there could be a way to detect this at page analysis (two mak:values
> having the same from&where within the same "level" of nesting, and if this case
> is detected, we could print a warning to the programmer.

yeh, there are plenty of warnings that could be printed (see bug 39) but where?
the result that the programmer gets is the page itself. should makumba send some
special http headers if the JSP is recent? one would need a proxy to interpret
those. maybe makumba should simply send back 2 frames, the makumba frame and the
result frame... another idea is to combine "makumba comments" with the page
output...

> on the other hand, if you'd implement a mak:stat tag, you could maybe provide
> sum, avg and count to be accessible by default.
> so, maybe something like this:
> <mak:stat field=".." from="....." where... /> // would do the query, and store
> sum(field), avg(field), count(field)

i thought of this, with the reserve that it's an expr, not a field so you can
accomodate avg(a+b+x), _and_ with the reserve that they should evaluate lazily
(i.e. computing all possible stats of the expr can cost a lot of CPU which you
don't want to commit to). about lastCount() and count(), i actually like it, see
above.

but i wouldn't complicate things so much just because programmers do copy-paste.
i'd rather count on some form of warnings displayed, plus i'd introduce a
special chapter "the unexpected server workload due to copy-paste" in the
makumba training. near the chapter "avoid bug 54"

**Comment by ghost** [comment]
**Comment by @rudolfmayer on 5 Jan 2005 21:27 UTC**
> another issue here is that for easy (automatic or manual) code conversion
> mak:lastCount() could be extended to mean "count of last <mak:list > or last
> count(*)"


that could be an option. i wonder though, would the mak:lastCount always have
the same value as a count(*)?
 
>>then, i think that it should not be a primary target to optimise existing
>>counting-only mak:lists, it is easy enough to detect lists without a body, and
>>then they should be replaced, and maybe this can be deprecated and not supported
>>anymore in the future.
> 
> 
> you can imagine that it's easier for me to let others do a replacement in JSP
> source instead of implementing new features of mak:list. however, without page
> analysis, that replacement cannot be automatic, due to the groupBy thing.


i did not mean automatic detection and replacement by makumba, rather manually
detecting these thigns with a grep and a regular expression, and then we
manually could change this. i really see it as an overkill to try to optimise
existing code, i would keep the focus on improved features, and then if the "end
user" (may it be ITD) does not take the effort for updating, then their fault.
 
> (here are a bunch of other replacements that need be done in karamba source to
> take advantage of new features, i made a list of these from 0.5.5 to 0.5.11,
> sent to mak-user@  July 2004, maybe we should update it)


i think that by now, i understand enough about this list, i'd like to transform
it into a FAQ somewhere on makumba.org, if it is available online it is a much
better help than in an email.
i could also add the stuff for bug 59/60 and the mak:list limit. if more
features are new, just pls give tell me which ones.
 
> "<mak:list /> is illegal. if the JSP file is more than 10 days old, honour it
> (maybe with the count() optimization). If the JSP file is recently developed,
> show an error and indicate usage of <mak:list> <mak:value
> expr="count(*)"/><mak:list>". 
> 
> only problem i see with this is that if you change a <br> to a <p> in the page
> you're stuck with a count(*) and groupBy :))). another solution is to say
> 
> "<mak:list /> is illegal. if the JSP file is more than 10 days old, honour it
> (maybe with the count() optimization). If the JSP file is recently developed,
> transform automatically <mak:list> <mak:value expr="count(*)"/><mak:list> in the
> JSP file."


i'd go for the first option, cause i guess it is less effort for implementing. i
could repeat my opinion from above, and just to underline this, lately we are
doing quite much refactoring and cleaning of the JSP code, e.g. gwen cleaned
most of helpdesk, lode does&did some on minerva, i did some on other PA pages.

so, trust on developers doing the change when you tell them to do.

>>i guess there could be a way to detect this at page analysis (two mak:values
>>having the same from&where within the same "level" of nesting, and if this case
>>is detected, we could print a warning to the programmer.
> 
> 
> yeh, there are plenty of warnings that could be printed (see bug 39) but where?
> the result that the programmer gets is the page itself. should makumba send some
> special http headers if the JSP is recent? one would need a proxy to interpret
> those. maybe makumba should simply send back 2 frames, the makumba frame and the
> result frame... another idea is to combine "makumba comments" with the page
> output...


i think the last thing is the best, there could be warnings printed on the top
of the page before the actualy page as long as you are in a developing
environment, like we have this non-production warning. 
as i guess that this is quite a best-specfic solution, maybe a boolean flag in a
makumba properties file could be set whether to show warnings or not (should
imho be global and not per context). 
 
>>on the other hand, if you'd implement a mak:stat tag, you could maybe provide
>>sum, avg and count to be accessible by default.
>>so, maybe something like this:
>><mak:stat field=".." from="....." where... /> // would do the query, and store
>>sum(field), avg(field), count(field)
> 
> 
> i thought of this, with the reserve that it's an expr, not a field so you can
> accomodate avg(a+b+x), _and_ with the reserve that they should evaluate lazily
> (i.e. computing all possible stats of the expr can cost a lot of CPU which you
> don't want to commit to). about lastCount() and count(), i actually like it, see
> above.


i am not exactly sure if i understand this, how it would look like. could you
give an example?
 
> but i wouldn't complicate things so much just because programmers do copy-paste.
> i'd rather count on some form of warnings displayed, plus i'd introduce a
> special chapter "the unexpected server workload due to copy-paste" in the
> makumba training. near the chapter "avoid bug 54"


fine for me too, as said above :-)

**Comment by ghost** [comment]
**Comment by @cristianbogdan on 5 Jan 2005 22:05 UTC**
> that could be an option. i wonder though, would the mak:lastCount always have
> the same value as a count(*)?

no, it will have the same value as the last count() of whatever was counted

>> special http headers if the JSP is recent? one would need a proxy to interpret
>> those. maybe makumba should simply send back 2 frames, the makumba frame and the
>> result frame... another idea is to combine "makumba comments" with the page
>> output...
> 
> 
> i think the last thing is the best, there could be warnings printed on the top
> of the page before the actualy page as long as you are in a developing
> environment, like we have this non-production warning. 

well, putting warnings before or after a page is one thing, and it's doable.
what i meant is putting stuff inside the produced page, at the respective place,
which is a bit more difficult.

maybe the current page that prints non-prod warning can automatically print a
makumba-specific header. then makumba won't have to alter the HTML, just set
that header. i'm not sure.

> i am not exactly sure if i understand this, how it would look like. could you
> give an example?

<mak:stat expr="x" ..../> ${mak:lastAvg() } 

would compute the average only when encountering mak:lastAvg() and would compute
nothing before. but i'm starting to doubt this solution more and more, as it's
pretty hard to implement and you agree that code copying should not affect mak
design

**Comment by ghost** [comment]
**Comment by @cristianbogdan on 26 Oct 2007 18:52 UTC**
i am amazed how complicated this bug became

i think the main complication is doing the counting in sub-mak:lists. i propose to simply file that as a separate bug and to implement something for this one so a count(*) will be selected at least in root mak:lists. that would help pagination (bug 990)

**Comment by ghost** [comment]
**Modified by @manuelbernhardt on 27 Jun 2010 19:42 UTC**

**Comment by ghost** [comment]
**Modified by @manuelbernhardt on 27 Jun 2010 19:43 UTC**

**Comment by ghost** [comment]
**Modified by @manuelbernhardt on 28 Jun 2010 07:36 UTC**
